### PR TITLE
Search History suggestions fixes #803

### LINKED ISSUE #803: Search History suggestions
From #775.

> Keeping history of searches in the LocalStorage of the user and allow them to remove them (GitLab does something like this IIRC, even Gmail).

This would mean that common searches would be auto-suggested.

### PR DESCRIPTION
**Pull request by prashuchaudhary**
* Added Functionality to Save Searches in Indexed DB.
* Wrote JS module to handle Search Suggestion
* Handled Search Suggestion Box HTML to be based on theme.

### CODE REVIEW

**Comment by prashuchaudhary** [comment]
Fixes #803

**Comment by andrew** [comment]
Great work! Do you have a screenshot?

**Comment by prashuchaudhary** [comment]
Search Suggestion ScreenShot

screen shot 2018-09-18 at 11 33 46 pm

File: app/assets/stylesheets/_application.scss [hunk 1]
```diff
@@ -744,3 +744,49 @@ svg{
...
+
+.search-dropdown {
+  top: 40px;
+  font-size: 14px;
+  padding: 0px;
+  left: 65px;
+  max-height: 300px;
+  overflow-y: auto;
+  flex-flow: column;
+  background: #fff;
```
[the remark this answers is not in the document]
Comment (prashuchaudhary) [reply app/assets/stylesheets/_application.scss]: will update this.

File: app/assets/stylesheets/_application.scss [hunk 2]
```diff
@@ -744,3 +744,49 @@ svg{
...
+}
+
+.search-remove-btn, search-remove-btn:hover {
+  font-weight: 400;
+  font-size: 12px;
+  text-decoration: none !important;
+  color: #000 !important;
+}
+
+.form-inline__item {
```
[the remark this answers is not in the document]
Comment (prashuchaudhary) [reply app/assets/stylesheets/_application.scss]: sure. will update

File: app/assets/stylesheets/_application.scss [hunk 3]
```diff
@@ -744,3 +744,49 @@ svg{
 .text-large{
   font-size: 16px;
 }
+
+.d-flex {
```
[the remark this answers is not in the document]
Comment (prashuchaudhary) [reply app/assets/stylesheets/_application.scss]: i didn't know. So I will check bootstrap 4 for all the default classes and will update code. Also will go through our `application.scss`

File: app/assets/javascripts/search_suggestion.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
+// create an instance of a db object for us to store the IDB data in
```
[the remark this answers is not in the document]
Comment (prashuchaudhary) [reply app/assets/javascripts/search_suggestion.js]: Sure. I will wrap the functionality in Modules.

**Comment by andrew** [comment]
Looking great, I just need to test this out on mobile

File: app/assets/javascripts/application.js [hunk 5]
```diff
@@ -412,10 +418,26 @@ function recoverPreviousCursorPosition() {
...
+  });
+
+  $("#search-box").on("click", function() {
+    SearchSuggestion.displaySearchSuggestions();
+  });
+
+  $("#search-sugguestion-list").on("click", function(event) {
+    SearchSuggestion.addToSearchBox(event);
+  });
+
```
[the remark this answers is not in the document]
Comment (prashuchaudhary) [reply app/assets/javascripts/application.js]: okk. doing that.

**Comment by andrew** [comment]
@prashuchaudhary few conflicts to solve here now that I've merged https://github.com/octobox/octobox/pull/951

**Comment by prashuchaudhary** [comment]
@andrew resolved merge conflict. updated the branch

**Comment by prashuchaudhary** [comment]
@andrew testing all the functionality again. wait for confirmation

**Comment by prashuchaudhary** [comment]
@andrew tested on Firefox, Chrome and Safari. Search working as intended. However, there is an issue with Search Suggestion Box size.

On Firefox, the Search Suggestion Box size matches the `INPUT` size, but on Chrome and Safari. It exceeds the `INPUT` box size.

**Comment by andrew** [comment]
Great, I'll check it out this afternoon 👍

**Comment by prashuchaudhary** [comment]
@andrew great

**Comment by prashuchaudhary** [comment]
@andrew fixed the requested changes

[x] remove the rounded corners on the dropdown 
[x] ensure the border on there is a border on the last history item
[x] help icon escapes the search box on firefox

**Comment by prashuchaudhary** [comment]
@andrew  I did not use octobox icon because i am creating the list Dynamic in JS. So octobox is not available. 
Either i have to create a variable in a ERB file OR copy HTML for octobox icon and then use it.

What do you suggest ??

**Comment by prashuchaudhary** [comment]
@andrew using Octobox icon as suggestion. Please review final changes.

**Comment by prashuchaudhary** [comment]
@andrew can you please do the final review today, all the requested changes done.

**Comment by andrew** [comment]
Yep will definitely review this today 👍

**Comment by prashuchaudhary** [comment]
thanks

**Comment by andrew** [comment]
Just one small visual change, can you line up the X with the ?:

screen shot 2018-09-25 at 14 02 51

Otherwise looking and working great 👍

**Comment by prashuchaudhary** [comment]
> Just one small visual change, can you line up the X with the ?:
> 
> screen shot 2018-09-25 at 14 02 51
> 
> Otherwise looking and working great 👍

Updated.

**Comment by andrew** [comment]
👌

**Pull request merged**